Drop support for EOL Pythons

jdufresne opened the pull request:
Support matrix now matches the test matrix from 6233f07bf303d442b39ac677fa6079acc64834a6. By matching these, helps to ensure all supported Python do indeed work.

Python 2.6 and 3.3 are end of life. They are no longer receiving bug fixes, including for security issues. Python 2.6 went EOL on 2013-10-29 and on 2017-09-29. For additional details on support Python versions, see:

Supported: https://devguide.python.org/#status-of-python-branches
EOL: https://devguide.python.org/devcycle/#end-of-life-branches

Removing support for EOL Pythons will reduce testing and maintenance resources while allowing the library to move towards a modern Python 3 style.

Using pypinfo, we can show the PyPI download statistics, show very low numbers for EOL Pythons.

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 2.7            |  48.94% |         37,036 |
| 3.6            |  26.58% |         20,113 |
| 3.5            |  15.38% |         11,642 |
| 3.4            |   5.93% |          4,488 |
| 3.7            |   3.13% |          2,365 |
| 3.3            |   0.03% |             21 |
| 2.6            |   0.01% |              9 |
| Total          |         |         75,674 |

jdufresne commented on natsort/ns_enum.py:
```diff
@@ -7,8 +7,6 @@
 
 import collections
 
-# NOTE: OrderedDict is not used below for compatibility with Python 2.6.
```
I'm happy to make this change, but I'm not sure what that should be. Can you clarify what the code change is?

SethMMorton replied:
Sorry about being cryptic.  I am specifically referring to `enum_fields` in the below code. It is written awkwardly because of the lack of `OrderedDict`. If `enum_fields` were an `OrderedDict` it wouldn't have to constantly be cast to `current_mapping`.

```python
# Construct the list of bitwise distinct enums with their fields.
enum_fields = [(name, 1 << i) for i, name in enumerate(enum_options)]
enum_fields.extend((name, 0) for name in enum_do_nothing)

for name, combo in enum_combos:
    current_mapping = dict(enum_fields)
    combined_value = current_mapping[combo[0]]
    for combo_name in combo[1:]:
        combined_value |= current_mapping[combo_name]
    enum_fields.append((name, combined_value))

current_mapping = dict(enum_fields)
enum_fields.extend((alias, current_mapping[name]) for alias, name in enum_aliases)

# Finally, extract out the enum field names and their values.
enum_field_names, enum_field_values = zip(*enum_fields)
```

jdufresne commented:
I've made the following changes:

- Simplified `natsort/ns_enum.py`.
- Added `test_ns_enum.py` to help ensure the final result stays the same. The values were a dump from the `master` branch.
- Simplified `natsort/__init__.py`

Please let me know if you see additional opportunities for simplifications in the code. I'm happy to make those changes. :slightly_smiling_face:

SethMMorton commented:
The last question about this PR is how to bump the version. The `python_requires` will prevent 2.6 users from accidentally installing a broken version, so maybe just bumping the minor version is OK. But on the other hand it is technically backwards-incompatible with 2.6 so perhaps a major version bump should be done?

jdufresne commented:
Ultimately, that is up to you as the package maintainer and I'll be fine either way. But I think if you have any concern you should bump the major version. It will best signal to package users that there are breaking changes. If you have previously considered any other backwards incompatible changes, now would be the time to get them in.

Rebased to resolve merge conflicts after the doc move.

SethMMorton commented:
Hmm, good suggestion.  There ARE a few other backwards incompatible changes I've been wanting to make, mostly removing old APIs.

Thanks for all your help!

SethMMorton commented:
So here is my plan of action:

 - I will release version 5.4.0 that will raise deprecation warnings for the API I will remove. This release will *not* include the changes from this merge.
 - In a few months I will release version 6.0.0, which will contain the contents of this merge. This will give users time to prepare for the API changes.

jdufresne commented:
sounds good thanks
